gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java

```diff
@@ -0,0 +1,470 @@
+import io.trino.sql.tree.ShowColumns;
+import io.trino.sql.tree.ShowCreate;
+import io.trino.sql.tree.ShowSchemas;
+import io.trino.sql.tree.ShowTables;
+import io.trino.sql.tree.Statement;
+import io.trino.sql.tree.Table;
+import io.trino.sql.tree.TableFunctionInvocation;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.ws.rs.HttpMethod;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.net.URLDecoder;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.InputMismatchException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.io.BaseEncoding.base64Url;
+import static io.airlift.json.JsonCodec.jsonCodec;
+import static java.lang.Math.toIntExact;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+import static java.util.Objects.requireNonNullElse;
+
+public class TrinoQueryProperties
+{
+    public static final String UNSET = "__UNSET__";
+
+    private final Logger log = Logger.get(TrinoQueryProperties.class);
+    private final boolean isClientsUseV2Format;
+    private String body = "";
+    private String queryType = "";
+    private Set<QualifiedName> tables = ImmutableSet.of();
+    private final String defaultCatalog;
+    private final String defaultSchema;
+    private Set<String> catalogs = ImmutableSet.of();
+    private Set<String> schemas = ImmutableSet.of();
+    private Set<String> catalogSchemas = ImmutableSet.of();
+    private boolean isQuery;
+
+    public static final String TRINO_CATALOG_HEADER_NAME = "X-Trino-Catalog";
+    public static final String TRINO_SCHEMA_HEADER_NAME = "X-Trino-Schema";
+    public static final String TRINO_PREPARED_STATEMENT_HEADER_NAME = "X-Trino-Prepared-Statement";
+
+    public TrinoQueryProperties(HttpServletRequest request, RequestAnalyzerConfig config)
+    {
+        isClientsUseV2Format = config.isClientsUseV2Format();
+
+        defaultCatalog = requireNonNullElse(request.getHeader(TRINO_CATALOG_HEADER_NAME), UNSET);
+        defaultSchema = requireNonNullElse(request.getHeader(TRINO_SCHEMA_HEADER_NAME), UNSET);
+        if (request.getMethod().equals(HttpMethod.POST)) {
+            processRequestBody(request, config);
+            isQuery = true;
+        }
+    }
+
+    private void processRequestBody(HttpServletRequest request, RequestAnalyzerConfig config)
+    {
+        try (BufferedReader reader = request.getReader()) {
+            if (reader == null) {
+                log.warn("HTTP request returned null reader");
+                body = "";
+                return;
+            }
+
+            Map<String, String> preparedStatements = getPreparedStatements(request);
+            SqlParser parser = new SqlParser();
+            reader.mark(config.getMaxBodySize());
+            body = CharStreams.toString(reader);
```

Comment:
the body size limit is there because, if i understand the semantics of `BufferedReader` correctly, reading is a destructive operation. I use the `mark` and `reset` methods to ensure I leave the `HttpServletRequest` in the same state that I received it.

If you read more than the`readAheadLimit` argument to `mark`, `reset` is not guaranteed to get you back to the mark. 

Prepared statements are request headers, which are plain Strings, so they don't need this handling.